<p><img src="https://files.cdn.thinkific.com/file_uploads/133292/images/a9e/c1a/bc3/P02.jpg"></p>

`MCSDSC02P01V3______`

`MCSDSC02P01V2______`

`MCSDSELIMC02P01V1010621`

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Спасибо, что доделал работу! Я оставил новые комментарии. По самому последнему этапу с модой, возможно, добавлю еще, но попозже — эту часть нужно получше формализовать в брифе сначала.

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Привет! Учитывая, что сейчас на курсе нет теории по предобратке на Pandas — твоя итерация очень достойная, ты крутой. Критических замечаний почти нет, но есть доработки.

Как обычно, заголовки (комментариев) будут трех цветов. <font color="89DD00">Зеленый</font> — все правильно (и если комментария нет, то это также значит, что все правильно). <font color="FFB800">Желтый</font> — можно доработать. <font color="FF0F00">Красный</font> — нужно исправить.

Постарайся учесть эти комментарии в ходе доработки проекта и не удаляй их.

<h3><font color="#3c00ff">Update</font></h3>

теорию я таки поизучал в курсе про ML, но конечно и погуглить много чего и пришлось) спасибо)

## Постановка задачи

#### Без очереди — это мобильное приложение, которое помогает сделать заказ до прихода в кафе, чтобы не ждать официанта, приготовление еды и счет.

Необходимо изучить данные сервиса с тестового запуска на территории дизайн-завода Флакон и Хлебозавода №9 и ответить на вопросы, поставленные менеджером проекта:

* На какие кафе приходится больше всего заказов? А на какие дни?
* Чаще ли приложением пользуются именно в обеденное время?
* Насколько большими компаниями  (или наоборот поодиночке) приходят в кафе?

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Норм

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

ОК, и еще бы сюда вставить предложение про сам сервис, что он делает.

<h3><font color="#3c00ff">Update</font></h3>

добавил

## Описание и план работ

### Шаг №1. Доступ к проекту

1. Создать аккаунт на GitHub и написать в чат, указав никнейм

2. Сделать клон репозитория на локальную машину

3. Перейти в директорию project_reviewing и создать в ней папку с никнеймом в телеграме.

4. Скопировать файл project_02_username.ipynb, подставив свой ник в телеграме, и перенести эту копию в ранее созданную папку

### Шаг №2. Получение данных

Импортировать нужные библиотеки и прочитать файл bez_ocheredi.csv из папки data. Получить первые 10 строк таблицы и общую информацию о наборе. Описать полученные данные.

### Шаг №3. Предобработка данных

Получить названия параметров (столбцов) и переименовать при обнаружении проблем.

Посчитать пропущенные значения. В строках с пропущенным названием ресторана заполнить его словом admin.

Посчитать полные дубликаты, удалить их и проверить, что больше их не осталось.

Вычислить дубликаты Iskra Bar и переименовать на правильное название.

### Шаг №4. Распределение заказов в течение недели

Посчитать, на какие кафе приходится больше всего заказов.

Сгруппировать данные по дню недели и посчитать, на какие дни приходится больше всего заказов.

Написать функцию, которая возвращает количество заказов для ресторана в определенный день.
Передать функции заведения Iskra Bar и Там Ем, и понедельник, среду и пятницу в качестве дня недели.
Свести результат в таблицу, в которой столбцами являются ресторан и 3 дня недели.

### Шаг №5. Распределение заказов в течение дня

Написать функцию, которая возвращает количество заказов за определенный промежуток времени.
Сравнить заказы с 12 до 16 и с 16 до 22.

### Шаг №6. Типичное количество человек за столиком

Написать функцию, которая определяет моду (типичное значение) количества человек, для которых заказывается столик. Протестировать функцию на Iskra Bar и Там Ем, передав ей дневной и вечерний промежуток времени.

### Шаг №7. Общие выводы

Написать общие выводы.

## Описание данных

* userID — уникальный ID пользователя
* ресторан — название ресторана
* weekday — день недели
* ordered — на какое время сделан заказ
* Столик для — для скольких людей заказан столик

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Норм

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Для построения ds-портфолио важны 2 вещи — блог, то есть умение рассказать о своем "ходе мышления" на проекте и код в репозиториях на GitHub. Формат ноутбука является 2 в 1 и сейчас вы описываете в нем свой проект для потенциального работодателя.

Я все это говорю к тому, что зачем ему видеть про «☑️ Готов к проверке» или «Критерии проверки»? Это инструкция/помощь в выполнении проекта от школы, можно опустить это)

<h3><font color="#3c00ff">Update</font></h3>

ок, убрал)

## Импорт данных и библиотек

In [1]:
import pandas as pd, os

Прочитаем файл `bez_ocheredi.csv` и сохраним его в переменную `bo_data`.

In [2]:
#data_dir = os.path.relpath('../../data')
#bo_data = pd.read_csv(os.path.join(data_dir, 'bez_ocheredi.csv'))

bo_data = pd.read_csv('/Users/elimilo/Desktop/mcs_ds_02/data/bez_ocheredi.csv')

Получим первые 10 строк таблицы.

In [1235]:
bo_data.head(10)

userID    ресторан     weekday  ordered    Столик для 
0  b912c3c6a6ad   Iskra Bar     Monday     10:00             1
1  2dac0150717d   Iskra Bar     Monday     13:30             2
2  ba83aefab5dc   Iskra Bar    Tuesday     12:30             3
3  02996f1dd2ea   Iskra Bar  Wednesday     12:00             2
4  7c90be56d2ab   Iskra Bar   Thursday     14:30             3
5  51e0962387f7      Там Ем     Friday     12:30             2
6  c4f260a2b48a         NaN   Saturday     15:00             6
7  6b71309d6a8a  Крафтейнер     Monday     10:30             1
8  c5b969cb63a2  Бар Сибирь  Wednesday     15:30             2
9  b1570962e68c  Бар Сибирь  Wednesday     12:30             3

Получим общую информацию о данных из таблицы.

In [1236]:
bo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1024 entries, 0 to 1023
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userID        1024 non-null   object
 1    ресторан     959 non-null    object
 2    weekday      1024 non-null   object
 3    ordered      1024 non-null   object
 4    Столик для   1024 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 40.1+ KB


In [1237]:
bo_data.describe(include='all')

userID   ресторан   weekday  ordered    Столик для 
count           1024        959     1024      1024   1024.000000
unique          1000          9        7        23           NaN
top     999d9913fe4d  Iskra Bar  Tuesday     14:30           NaN
freq               7        269      252       101           NaN
mean             NaN        NaN      NaN       NaN      1.752930
std              NaN        NaN      NaN       NaN      0.831028
min              NaN        NaN      NaN       NaN      1.000000
25%              NaN        NaN      NaN       NaN      1.000000
50%              NaN        NaN      NaN       NaN      2.000000
75%              NaN        NaN      NaN       NaN      2.000000
max              NaN        NaN      NaN       NaN      6.000000

In [1238]:
bo_data.nunique()

userID          1000
 ресторан          9
 weekday           7
 ordered          23
 Столик для        6
dtype: int64

In [1239]:
bo_data.shape

(1024, 5)

In [1240]:
bo_data.dtypes

userID          object
 ресторан       object
 weekday        object
 ordered        object
 Столик для      int64
dtype: object

Столбцы в `bo_data` содержат следующую информацию:

* userID — уникальный ID пользователя
* ресторан — название ресторана
* weekday — день недели
* ordered — на какое время сделан заказ
* Столик для — для скольких людей заказан столик

### Выводы

* в датасете 1024 строки и 5 столбцов.
* все столбцы, за исключением последнего, строковые. Последний столбец содержит целые числа от 1 до 6.
* в столбце userid 1000 уникальных значений при общем кол-ве строк 1024. Это свидетельствует о наличии дубликатов.
* столбец "ресторан" содержит пропущенные значения.
*  с днями недели, видимо, все ок. По крайней мере их 7))

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

ОК, но теперь чересчур подробно. `info()` самый оптимальный здесь, все так, про остальные я к слову упомянул)

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

По этой части несколько комментариев:

1. Передавай в `head()` побольше строк, иначе можно не увидеть структуру данных (особенно в следующих кейсах)


2. К методу `info()` добавлю, что можно еще использовать атрибут `dtypes` для более компактного вывода типов и атрибут `shape` для вывода размерности


3. По дефолту метод `describe()` возвращает только числовые типы, передавай тогда `include='all'`, а то сейчас не очень много пользы


4. Выводы OK — указываешь на структуру данных, а также обращаешь внимание на дубликаты и пропуски

<h3><font color="#3c00ff">Update</font></h3>

1. изменил `head(10)`
2. да, спасибо. Ну я не стал отдельно прописывать `shape` и `dtypes`, т.к. эта инфа уже есть в `info()`
3. о! так информативней конечно)
4. ок

## Предобработка данных

Получим названия столбцов.

In [3]:
bo_data.columns

Index(['userID ', ' ресторан ', ' weekday', ' ordered ', ' Столик для '], dtype='object')

Обращаем внимание, что помимо названий на разном языке, в них есть пробелы, которые затрудняют доступ к данным. Переименуем столбцы и проверим результат.

In [4]:
#bo_data = bo_data.rename(columns = {'userID ':'user_id',
#                          ' ресторан ':'restaurant',
#                         ' weekday':'weekday',
#                          ' ordered ':'ordered_time',
#                          ' Столик для ':'guests_per_table'})

bo_data.set_axis(['user_id', 'restaurant', 'weekday', 'ordered_time', 'guests_per_table'], axis=1, inplace=True)

In [5]:
bo_data.columns

Index(['user_id', 'restaurant', 'weekday', 'ordered_time', 'guests_per_table'], dtype='object')

Еще раз проверим данные на наличие пропусков, просуммировав их.

In [6]:
bo_data.isna().sum()

user_id              0
restaurant          65
weekday              0
ordered_time         0
guests_per_table     0
dtype: int64

Менеджер проекта предупреждал, что некоторые заказы заводились на тестовую карточку кафе без названия, поэтому есть смысл заполнить такой пропуск словом admin.

In [7]:
bo_data['restaurant'] = bo_data['restaurant'].fillna('admin')

Проверяем, что пропусков не осталось.

In [8]:
bo_data.isna().sum()

user_id             0
restaurant          0
weekday             0
ordered_time        0
guests_per_table    0
dtype: int64

Установим количество явных дубликатов. Если найдутся, то удалим и снова проверим, что их не осталось.

In [9]:
dup_quantity = bo_data.duplicated().sum()
bo_data.duplicated().sum()

24

In [22]:
bo_data = bo_data.drop_duplicates()
bo_data = bo_data.reset_index()

In [28]:
bo_data.loc[20:25]

index       user_id  restaurant    weekday ordered_time  guests_per_table
20     20  ea12aec32ec3     Северия    Tuesday        12:30                 4
21     21  838a82cec0a6       admin  Wednesday        13:00                 4
22     26  de310e6aae86     Северия   Thursday        12:00                 2
23     27  08b24e170109     Северия   Saturday        11:00                 3
24     28  065a4daba35f  Бар Сибирь    Tuesday        13:30                 1
25     29  990113ae4f1c  Крафтейнер     Monday        13:30                 1

<h3><font color="#FFB800">Комментарий ассистента 2</font></h3>

Вывел срез, в котором были дубликаты, чтобы показать, что если не передавать `drop=True` методу `reset_index()`, то он добавляет его как отдельный столбец. Нам такое не надо. 

И еще пиши цепочкой все-таки `drop_duplicates().reset_index(drop=True)`, а то зачем строки расходовать :-))

In [1249]:
bo_data.duplicated().sum()

0

Скорее всего, такие полные дубликаты говорят о сбое при выгрузке данных. Необходимо сообщить об этом.

In [1250]:
if dup_quantity > 0:
    print('В датасете обнаружено {} записей-дубликатов. Возможно, произошел сбой при выгрузке данных'.format(dup_quantity))

В датасете обнаружено 24 записей-дубликатов. Возможно, произошел сбой при выгрузке данных


Еще одна проблема, о которой предупреждал менеджер проекта — дубликаты карточки Iskra Bar с неправильными названием бара - "Искра" и "Искр". На них заводились реальные заказы, поэтому необходимо вычислить их и переименовать на правильное название — Iskra Bar.

Объявим функцию `replace_restaurant()`, которая заменяет неправильное название, а потом проверяет успешность выполнения замены подсчетом (если замена прошла успешно — результат подсчета должен быть 0).

In [31]:
def replace_restaurant(dataset, oldname,newname):
    '''берет на вход датасет, старое и новое название ресторанов и исправляет записи'''
    print('Количество записей с ошибкой в названии ресторана: {}'.format(len(dataset.loc[dataset['restaurant'] == oldname])))
    dataset.loc[dataset['restaurant'] == oldname, 'restaurant'] = newname
    print('Исправляем записи...')
    print('Количество записей с ошибкой в названии ресторана: {}'.format(len(dataset.loc[dataset['restaurant'] == oldname])))

In [32]:
def replace_restaurant_2(df, oldname, newname):
    df['restaurant'] = df['restaurant'].replace(oldname, newname)
    result = df[df['restaurant'] == oldname]['restaurant'].count()
    
    return result

Передаем функции названия 'Искра' и 'Искр'.

In [33]:
replace_restaurant_2(bo_data, 'Искра','Iskra Bar')

0

In [34]:
replace_restaurant_2(bo_data, 'Искр','Iskra Bar')

0

### Выводы

* наименования столбцов поправлены, но логично было бы исправить мэппинг еще на стороне приложения.
* некорректные названия ресторана Iskra bar также заменены, но логично также исправить ошибки на стороне приложения.

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

По третьему пункту уточнил выше, остальное ок.

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

1. Для переименования можно еще использовать метод `set_axis()` со следующими аргументами: список с названиями столбцов, `axis='columns'`, `inplace = True`.


2. `userID` — это camel case, а `guests_per_table` — snake_case. Лучше на чем-то одном остановиться и еще `ordered` сделать очевиднее: например, `ordered_time`.


3. Когда ищешь явные дубликаты, то одной строкой можно обойтись: `bo_data.duplicated().sum()`. Дальше тоже можно просто цепочку методов использовать: `bo_data = bo_data.drop_duplicates().reset_index(drop=True)`


4. Прикольное условие сделал, только под «сообщить» имелось в виду просто сказать разработчику, который выгружал для тебя данные :-))


5. Для `replace_restaurant()` — интересное решение! Посмотри еще вот такую реализацию, в ней можно одним методом `replace()` обойтись, а потом "проверить" его выполнение:

```
def replace_restaurant(df, wrong):
    df['restaurant'] = df['restaurant'].replace(wrong, 'Iskra Bar')
    result = df[df['restaurant'] == wrong]['restaurant'].count()
    
    return result
```

<h3><font color="#3c00ff">Update</font></h3>

1. попробовал.
2. остановился на snake_case
3. так я вроде и обошелся одной строкой `bo_data.duplicated().sum()`. Если ты про строку, где я завожу переменную, то это только для условия
4. я как раз долго думал, что тут имелось в виду)) решил, что лишним не будет)
5. попробовал, ок

## Распределение заказов в течение недели

Считаем, на какие кафе приходится больше всего заказов.

In [1255]:
bo_data['restaurant'].value_counts()

Iskra Bar     287
Там Ем        200
Бар Сибирь    159
Северия       107
Крафтейнер     78
Коворкафе      66
Азия           62
admin          41
Name: restaurant, dtype: int64

Группируем по `weekday` и считаем, на какие дни приходится больше всего заказов.

In [1256]:
bo_data.groupby('weekday')['restaurant'].count().sort_values(ascending=False)

weekday
Tuesday      251
Wednesday    242
Thursday     180
Monday       141
Friday       110
Saturday      52
Sunday        24
Name: restaurant, dtype: int64

In [1257]:
bo_data['weekday'].value_counts()

Tuesday      251
Wednesday    242
Thursday     180
Monday       141
Friday       110
Saturday      52
Sunday        24
Name: weekday, dtype: int64

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Да, со днями недели тоже можно разобраться с помощью `value_counts()` на самом деле, просто хотелось побольше группировки добавить. Думаю, поправим описание шага тут

Объявляем функцию, которая возвращает количество заказов для ресторана в определенный день.

In [35]:
def orders_per_weekday(dataset, restaurant, weekday):
    '''берет на вход датасет, ресторан и день недели,
    выдает кол-во заказов по запросу'''
    cond1 = dataset['restaurant'] == restaurant
    cond2 = dataset['weekday'] == weekday
    return dataset[cond1 & cond2]['restaurant'].count()

Передадим ей заведения Iskra Bar и Там Ем — два флагманских места, по которым просили результаты прежде всего — и понедельник, среду и пятницу в качестве дня недели.

In [36]:
orders_per_weekday(bo_data, 'Iskra Bar', 'Monday')

48

In [37]:
orders_per_weekday(bo_data, 'Iskra Bar', 'Wednesday')

53

In [38]:
orders_per_weekday(bo_data, 'Iskra Bar', 'Friday')

29

In [39]:
orders_per_weekday(bo_data, 'Там Ем', 'Monday')

25

In [40]:
orders_per_weekday(bo_data, 'Там Ем', 'Wednesday')

52

In [41]:
orders_per_weekday(bo_data, 'Там Ем', 'Friday')

24

Сведем результат в таблицу.

In [51]:
iskra_mon = orders_per_weekday(bo_data, 'Iskra Bar', 'Monday')
iskra_wed = orders_per_weekday(bo_data, 'Iskra Bar', 'Wednesday')
iskra_fri = orders_per_weekday(bo_data, 'Iskra Bar', 'Friday')
tam_mon = orders_per_weekday(bo_data, 'Там Ем', 'Monday')
tam_wed = orders_per_weekday(bo_data, 'Там Ем', 'Wednesday')
tam_fri = orders_per_weekday(bo_data, 'Там Ем', 'Friday')

In [78]:
data = [
    ['Iskra Bar', iskra_mon, iskra_wed, iskra_fri],
    ['Там Ем', tam_mon, tam_wed, tam_fri]
]

columns = ['restaurant', 'monday', 'wednesday', 'friday']

result = pd.DataFrame(data=data, columns=columns)
result

restaurant  monday  wednesday  friday
0  Iskra Bar      48         53      29
1     Там Ем      25         52      24

In [1267]:
#temp.pivot(index='restaurant', columns='weekday', values='orders_per_weekday')

### Выводы

* заведения с наибольшим кол-вом заказов - Искра бар и Там Ем
* основной поток заказов наблюдается в середине недели в будние дни (со вторника по четверг). Это довольно логично, в будние дни приходит много офисных работников на бизнес-ланчи.
* данные по заведениям Искра бар и Там Ем соответствуют общему тренду.

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

По третьему пункту твой вопрос — да, запись получается компактнее. А так ты заводишь отдельные переменные для каждой части условия.

По четвертому пункту — попробуй тогда передать конструктру список `index=['Iskra Bar', 'Там Ем']` и убрать соответствующий столбец restaurant.

И еще обрати внимание на `print()` — если его явно прописать, то вывод некрасивый. Можно только саму переменную указать, будет табличка

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

1. Когда считаем количество заказов для кафе, можно посчитать количество каждого ресторана просто: `bo_data['restaurant'].value_counts()`.


2. Сгруппировать по дням тоже можно немного полегче: `bo_data.groupby('weekday')['restaurant'].count().sort_values(ascending=False)`. В квадратных скобках — это значение какого показателя выводим, а в круглых — по какому параметру группируем.


3. В функции `orders_per_weekday()` можно использовать логическое условие: `restaurants_list = df[(df['weekday'] == day) & (df['restaurant'] == restaurant)]`. Внутри квадратных скобок указываем 2 условия, которые записываются в круглых, а между ними специальный оператор `&`, который работает как `and` в питоне. Второй строкой считаем кол-во ресторанов: `restaurants_list_count = restaurants_list['restaurant'].count()`.


4. У тебя получились правильные значения, но при сведении в таблицу они оказались не на тех местах (см. wednesday). Я бы предложил пока ограничиться конструктором, метод `pivot_table()` попробуем в следующем проекте. Поэтому можно просто сохранить результат функции, а потом сделать из этих значений табличку:

```
data = [
    ['Iskra Bar', iskra_mon, iskra_wed, iskra_fri],
    ['Там Ем', tam_mon, tam_wed, tam_fri]
]

columns = ['restaurant', 'monday', 'wednesday', 'friday']

result = pd.DataFrame(data=data, columns=columns)
print(table)
```

5. Про первый вывод — мы не можем утверждать, что гости Искры и Там Ем ходят к ним активнее, потому что эти заведения в целом могут принять большее количество гостей, нежели Азия или Коворкафе. Величины заполняемости могут быть сопоставимы здесь.

<h3><font color="#3c00ff">Update</font></h3>

1. так проще конечно! спасибо)
2. а тут получается тоже проще написать `bo_data['weekday'].value_counts()`? Результат ведь один и тот же.
3. заменил `len` на `count()`. Вопрос: чем твой вариант принципиально лучше? То, что все условия лучше сразу записывать в одну переменную?
4. заменил, спасибо. а можно как-то при выводе таблицы не выводить столбец с индексом?
5. под "популярными" я имел в виду заведения с наибольшим кол-вом заказов. заменил формулировку.

## Распределение заказов в течение дня

Сравним количества заказов, сделанных в обеденное и вечернее время — с 12 до 16 и с 16 до 22. Для этого объявим функцию, которая принимает данные и 2 аргумента времени.

In [1268]:
# время в столбце "ordered" приходит в строковом формате, необходимо поменять на формат времени
bo_data['ordered_time'] = pd.to_datetime(bo_data['ordered_time'], format='%H:%M')

In [1269]:
def spread_during_daytime(dataset, start_time, end_time):
    '''берет на вход датасет, начальное и конечное время (в формате HH:MM),
    выдает кол-во заказов за период'''
    
    condition = (dataset['ordered_time'] >= pd.to_datetime(start_time, format='%H:%M')) & (dataset['ordered_time'] <= pd.to_datetime(end_time, format='%H:%M'))
    return dataset[condition]['ordered_time'].count()

In [1270]:
def spread_during_daytime2(dataset, start_time, end_time):
    '''берет на вход датасет, начальное и конечное время (в формате HH:MM),
    выдает кол-во заказов за период'''
    
    condition = (dataset['ordered_time'] >= start_time) & (dataset['ordered_time'] <= end_time)
    return dataset[condition]['ordered_time'].count()

Передадим промежуток с 12:00 и 16:00

In [1271]:
spread_during_daytime(bo_data, '12:00', '16:00')

588

In [1272]:
#spread_during_daytime2(bo_data, '12:00', '16:00')

И с 16:00 до 22:00

In [1273]:
spread_during_daytime(bo_data, '16:00', '22:00')

338

In [1274]:
spread_during_daytime(bo_data, '06:00', '12:00')

175

### Выводы

* больше половины заказов приходится на дневное время, треть - на вечернее, и незначительное кол-во заказов (около 10%) на утро.
* гипотеза: утром и днем люди делают недорогие заказы(завтраки, бизнес-ланчи, кофе,чай), а вечером могут заказывать более дорогие блюда и алкоголь. Это может повлиять на средний чек. Целесообразно в дальнейшем передавать датасет также с данными по стоимости заказа для анализа.

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Заказ, сделанный в 16:00 — уже к вечернему времени относится, ты прав. My bad)

<h3><font color="FFB800">Комментарий ассистента</font></h3>

Большой респект, что сам догадался конвертировать строку во время! Мы этого даже не закладывали в этот проект (в следующий — да). Я тебе поправил знак, чтобы он учитывал крайнее время, а также рекомендую использовать и здесь логическое условие `df[(df['column1'] == value1) & (df['column2'] == value2)]`, будет изящнее. Также лучше использовать метод pandas `count()`, а не `len()` — прицепляй его в конце к логическому условию.

Кстати, если бы ты не конвертировал и сравнивал прямо строки — результат тоже был бы правильный.

Вывод здесь главный в том, что большинство заказов делают именно в обед. Это то, что и просили от нас узнать на текущем этапе.

<h3><font color="#3c00ff">Update</font></h3>

> Я тебе поправил знак, чтобы он учитывал крайнее время

разве это правильно? Получается, что так дважды учитываются заказы с пограничным временем: 588+388+175=1101, а в моем варианте 543+338+119=1000 (всего в датасете как раз 1000 строк)

> рекомендую использовать и здесь логическое условие

ок, заменил

> Кстати, если бы ты не конвертировал и сравнивал прямо строки — результат тоже был бы правильный.

не понимаю почему, но когда я писал условие только с одним временем, то у меня все работало и со строкой. а когда указывал два времени, то выводилась ошибка. А сейчас попробовал, и все получилось)) видимо при объединении где-то ошибка была)

## Типичное количество человек за столиком

Выясним, для какого количества человек обычно заказывают столик в разное время на примере Iskra Bar и Там Ем. Для этого напишем функцию, которая принимает данные, название ресторана, 2 времени и считает моду.

In [91]:
def guests_per_table(dataset, restaurant, start_time, end_time):
    '''берет на вход датасет, ресторан, начальное и конечное время (в формате HH:MM),
    выдает кол-во заказов за период'''
    
    condition = ((dataset['ordered_time'] >= pd.to_datetime(start_time, format='%H:%M')) &
                (dataset['ordered_time'] < pd.to_datetime(end_time, format='%H:%M')) &
                (dataset['restaurant'] == restaurant))
    temp = dataset[condition]
    print(temp['guests_per_table'].value_counts())
    print('\nЧаще всего с {} до {} часов в ресторане {} столик заказывают для {} чел.'.format(start_time, end_time,restaurant, int(temp['guests_per_table'].mode())))

Проверим функцию на двух ресторанах, а также дневном и вечернем промежутке времени.

In [92]:
guests_per_table(bo_data, 'Iskra Bar', '12:00', '16:00')

2    72
1    67
3    19
4     2
5     1
Name: guests_per_table, dtype: int64

Чаще всего с 12:00 до 16:00 часов в ресторане Iskra Bar столик заказывают для 2 чел.


In [1277]:
guests_per_table(bo_data, 'Iskra Bar', '16:00', '22:00')

1    45
2    35
3    17
4     1
Name: guests_per_table, dtype: int64

Чаще всего с 16:00 до 22:00 часов в ресторане Iskra Bar столик заказывают для 1 чел.


In [1278]:
guests_per_table(bo_data, 'Там Ем', '12:00', '16:00')

2    43
1    35
3    15
4     3
6     1
Name: guests_per_table, dtype: int64

Чаще всего с 12:00 до 16:00 часов в ресторане Там Ем столик заказывают для 2 чел.


In [1279]:
guests_per_table(bo_data, 'Там Ем', '16:00', '22:00')

1    34
2    31
3    11
6     2
Name: guests_per_table, dtype: int64

Чаще всего с 16:00 до 22:00 часов в ресторане Там Ем столик заказывают для 1 чел.


### Выводы

* Как в обеденное, так и в вечернее время, чаще всего столик заказывают для 1-2 чел.
* возможно стоит подумать над какой-нибудь маркетинговой акцией, если заказываешь через приложение столик для большого кол-ва гостей (чем больше гостей, тем больше скидка и т.д.)

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Ок, тогда мы в брифе формализуем получше как эту часть проекта решить, у других тоже вопросы возникли. Как минимум, укажем, что можно пока без преобразования обойтись. Если мне будет что добавить к этому этапу, то я прямо «Комментарий 3» заведу и дам знать тебе.

<h3><font color="FFB800">Комментарий ассистента</font></h3>

Ты получаешь правильный результат, очень крутые аутпуты-комментарии, но как и выше я рекомендую использовать логическое условие с методом `mode()` на конце.

Вывод правильный, люди используют приложение для 1-2 человек и практически не делают предзаказ для больших компаний. Это то, что от нас просили.

<h3><font color="#3c00ff">Update</font></h3>

так я и использую вроде в последней строке метод `mode()`.  

А `print(temp['guests_per_table'].value_counts())` я применил для того, чтобы было видно, что разница между заказами на 1 и 2 чел. небольшая.

## Общие выводы

* целесообразно поправить процедуру выгрузки входящих данных
* заведения с наибольшим кол-вом заказов - Искра бар и Там Ем
* основной поток заказов наблюдается в середине недели в будние дни (со вторника по четверг). Это довольно логично, в будние дни приходит много офисных работников на бизнес-ланчи.
* больше половины заказов приходится на дневное время, треть - на вечернее, и незначительное кол-во заказов (около 10%) на утро.
* гипотеза: утром и днем люди делают недорогие заказы(завтраки, бизнес-ланчи, кофе,чай), а вечером могут заказывать более дорогие блюда и алкоголь. Это может повлиять на средний чек. Целесообразно в дальнейшем передавать датасет также с данными по стоимости заказа для анализа.
* Как в обеденное, так и в вечернее время, чаще всего столик заказывают для 1-2 чел.
* возможно стоит подумать над какой-нибудь маркетинговой акцией, если заказываешь через приложение столик для большого кол-ва гостей (чем больше гостей, тем больше скидка и т.д.)

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Теперь отлично.

<h3><font color="FFB800">Комментарий ассистента</font></h3>

Все верно, только обрати еще раз внимание на второй вывод. Азия может иметь 60 посадочных мест и заполнять их на 100%, а Искра 500 и только половину. Если сравнивать так, то Азия популярнее.

<h3><font color="#3c00ff">Update</font></h3>

поправил!

Спасибо за комментарии)